homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,240 @@
+"""Module to coordinate llm tools."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Callable, Iterable
+from dataclasses import dataclass
+from functools import cached_property
+import logging
+from types import FunctionType
+from typing import Any
+
+import voluptuous as vol
+from voluptuous_openapi import UNSUPPORTED, convert
+
+from homeassistant.core import Context, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+
+from . import area_registry as ar, config_validation as cv, floor_registry as fr, intent
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_KEY = "llm_tool"
+
+
+def default_custom_serializer(schema: Any) -> Any:
+    """Serialize additional types in OpenAPI-compatible format."""
+    from homeassistant.util.color import (  # pylint: disable=import-outside-toplevel
+        color_name_to_rgb,
+    )
+
+    if schema is cv.string:
+        return {"type": "string"}
+
+    if schema is cv.boolean:
+        return {"type": "boolean"}
+
+    if schema is color_name_to_rgb:
+        return {"type": "string"}
+
+    if isinstance(schema, cv.multi_select):
+        return {"enum": schema.options}
+
+    if isinstance(schema, FunctionType):
+        return {}
+
+    return UNSUPPORTED
+
+
+@dataclass(slots=True)
+class ToolInput:
+    """Tool input to be processed."""
+
+    tool_name: str
+    tool_args: dict[str, Any]
+    platform: str
+    context: Context | None
+    user_prompt: str | None
+    language: str | None
+    agent_id: str | None
+    conversation_id: str | None
+    device_id: str | None
+    assistant: str | None
+
+
+class Tool:
+    """LLM Tool base class."""
+
+    name: str
+    description: str
+    parameters: vol.Schema
+    custom_serializer: Callable[[Any], Any]
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        parameters: vol.Schema = vol.Schema({}),
+        custom_serializer: Callable[[Any], Any] = default_custom_serializer,
+    ) -> None:
+        """Init the class."""
+        self.name = name
+        self.description = description
+        self.parameters = parameters
+        self.custom_serializer = custom_serializer
+
+    @cached_property
+    def specification(self) -> dict[str, Any]:
+        """Get the tool specification in OpenAPI-compatible format."""
+        return {
+            "name": self.name,
+            "description": self.description,
+            "parameters": convert(
+                self.parameters, custom_serializer=self.custom_serializer
+            ),
+        }
+
+    @abstractmethod
+    async def async_call(self, hass: HomeAssistant, tool_input: ToolInput) -> Any:
+        """Call the tool."""
+        raise NotImplementedError
+
+    def as_dict(self) -> dict[str, Any]:
+        """Get the tool specification in OpenAPI-compatible format."""
+        return self.specification
+
```

Comment:
Right now we have 3 LLM integrations: [OpenAI](https://www.home-assistant.io/integrations/openai_conversation), [Google](https://www.home-assistant.io/integrations/google_generative_ai_conversation/) and [Ollama](https://rc.home-assistant.io/integrations/ollama).
Ollama does not currently support function calling, thought there could be workarounds.
Both OpenAI and Google use OpenAPI-based function specifications, and I would expect this to be an industry standard.
Technically, Google Python API uses protobuf-based API, but it is still based on OpenAPI, so it looks like that probably all LLM integrations will use OpenAPI format.
We can move the logic to the integrations, and that would also add a benefit of moving the dependency of `voluptuous-openapi` from core to integrations, but my idea was that if every integration would need it, then it is better be put in the helpers.
But if my idea is wrong, I will move it.

Reply:
I could see other LLMs just using their own arbitrary formats, json, whatever. I think we should not over generalize this yet even though it means there is duplicate code in two integrations. The library helps a ton with that!

Reply:
Could you give some examples of the other formats?

Reply:
Even if explaining it all to LLM in the prompt manually (as we have to do for Ollama), I still believe the best way to explain the format to the LLM is OpenAPI (also known to LLMs as `json schema`).
This is not a contradiction to the architecture discussion, it would still be the responsibility of the LLM integration to expose the API to the LLM, even if we provide the helper of translating `vol.Schema` into OpenAPI schema.

Let me summarize the discussion about keeping the OpenAPI schema in the helpers.

Pros:
1. It is likely that every LLM integration will want to use it as a first step (while it doesn't have to)
2. It provides more flexibility to the LLM tool developer
3. Keeps the LLM integration code more clean
4. Improves maintainability by reducing code duplication

Cons:
1. Keeps the helpers code less clean
2. Decreases the responsibility separation
3. Adds `voluptuous-openapi` dependency on core level

I hope I captured all of it. I do not have any more arguments, so if you are still certain, I will remove it.

Reply:
I don't think that captures it.

My point is that given you went to the effort to make a library, then it should be very straight forward to call it from the integration such that having `voluptious_openai` with a nicely defined API is even better than having logic in `llm.py` here since you can have all of the wins in your list above for both the pros and cons.  The integration code is straight forward:

```
tools = llm.async_get_tools()
tool_specs = [
    voluptious_openai.convert(tool.name, tool.description, tool.parameters)
    for tool in tools
]
```

IMO that is clean, flexible, easy to reuse, keeps the helper code code, improves responsibility separation, and puts the dependencies where they should be.

Reply:
While I made my best effort to cover as many cases as possible, the conversion is not one to one and will never be.
Now, if a tool developer wants to tweak the conversion of his schema into open API, instead of just defining the `custom_serializer` method in his own class, he would need to do it every LLM conversation agent, including custom ones. If his case is not general enough, it wouldn't be possible to do that in the library, that's what the `custom_serializer` is for. So, effectively, we are just limiting the developer to the features that we can convert already.

Also, the result of the conversion is not that nicely cached within the class itself as with `@cached_property`.

Perhaps there is some confusion about the naming, It is `Open API`, not `OpenAI`. Open API is an open standard and is not vendor-specific.

Reply:
Sorry that was a typo, I understand it's openapi and not openai.

Is there a specific type you are worried about converting here right now? My additional worry is also that it's premature as I'm not sure what the concrete use case is that is needed here.

It also seems possible for the tool developer to handle this by changing the schema to be something more universal (e.g. strings) rather than some complex type it needed.

Caching can be added to the library if that is important.

Reply:
No, I have no specific example. I also agree with you it might be premature. Let's see. We can always add it later.

Reply:
Yes, 100% happy to add back as the use cases arise.